rfcs/0000_hypotheses_resampling.md

```diff
@@ -0,0 +1,75 @@
+- Start Date: 2025-02-27
+- RFC PR: (leave this empty, it will be filled in after RFC is merged)
+
+# Summary
+
+Resample hypotheses at every step in a manner inspired by particle-filters. This is the first step for Monty to interact with multiple objects and recognize compositional objects.
+
+# High-Level Motivation
+
+In an unsupervised learning experiment setup, Monty may be presented with multiple objects in a single episode. Ideally, we would like to move away from the traditional data loading setup of machine learning where there is a strict definition of an epoch, episode and step. As Monty starts to interact with the real world, the definition of epoch and episode will start to fade away and we'll be left with simple time discretization (i.e., step). The current definitions are:
+* Epoch: Used by the experiment class to denote one full pass through all the objects
+* Episode: Denotes a change in object
+* Step:  Denotes a single sensation and action in the sensorimotor framework.
+
+Real world interactions do not have epochs or episodes (these are only used for performance benchmarks), instead we could imagine the agent wondering around in a multi-object dynamic environment. The objects can be occluded, moving, or even disappearing behind new objects. 
+
+**We want Monty to handle dynamic environments by seamlessly switching from one object to another as it's sensors move around on the different objects.**
+
+# The Problem
+Monty is designed to receive a weak supervision signal when the episode ends and a new episode begins (change of object). This signal performs a full reset of all states within Monty. This reset includes counters, buffer, goal state generators, learning modules and sensory modules. Additionally, this reset sets Monty back into Matching mode. The below figure shows where this resetting is done. Most resetting happens in the `pre_episode` functions of the Monty and SMs and LMs classes.
```

Comment:
*a weak supervision signal --> a weak supervision signal at inference time

Clarifying because this weak supervision will still exist at learning after this RFC.